File: solutions/386/main.py
from typing import * 

class Solution:
    def traverse(self, start, n):
        result = [] 
        for i in range(10): 
            j = start*10 + i 
            if j > n: 
                return result 
            # visit 
            result += [j] 
            # L, R 
            result += self.traverse(j, n)
        return result 
    def lexicalOrder(self, n: int) -> List[int]:
        result = []
        if n < 10: 
            return list(range(1, n+1))
        for i in range(1, 10): 
            # visit 
            result += [i]
            # L, R
            result += self.traverse(i, n)
        return result 

if __name__ == "__main__": 
    solution = Solution()

    n = 13 
    result = solution.lexicalOrder(n) 
    assert result == [1, 10, 11, 12, 13, 2, 3, 4, 5, 6, 7, 8, 9] 

File: solutions/386/main.rs
impl Solution {
    fn pre_order(start_idx: i32, n: i32) -> Vec<i32> {
        let mut res = Vec::new(); 
        for i in 0..10 {
            let j = start_idx*10 + i; 
            // base case 
            if j > n {
                return res; 
            }
            // visit 
            res.push(j); 
            // traverse 
            res.extend(Solution::pre_order(j, n));
        }
        res 
    }
    pub fn lexical_order(n: i32) -> Vec<i32> {
        let mut res = Vec::new(); 
        if n < 10 {
            for i in 1..(n+1) {
                res.push(i); 
            }
            return res; 
        }
        for i in 1..10 {
            // visit 
            res.push(i); 
            // traverse 
            res.extend(Solution::pre_order(i, n)); 
        }
        res 
    }
}

struct Solution {}

fn main() {
    assert_eq!(vec![1,10,11,12,13,2,3,4,5,6,7,8,9], Solution::lexical_order(13)); 
}